### PR TITLE
Support for log filter policies via hooks, deprecation of the existing predicate functions

### PR DESCRIPTION
**Pull request by ckreibich**
This deprecates the `$pred` functions on `Log::Filter` instances and establishes the use of hooks as a way to express "policies" on log writes going through filters. The idea is as follows:

We standardize a hook for acting on log records:

```
module Log;

export {
    ...
    
    ## A hook type to implement filtering policy. Hook handlers can
    ## veto the logging of a record or alter it prior to logging.
    ## You can pass arbitrary state into the hook via the
    ## filter argument and its config member.
    ##
    ## rec: An instance of the stream's ``columns`` type with its
    ##      fields set to the values to be logged.
    ##
    ## id: The ID associated with the logging stream the filter
    ##     belongs to.
    ##
    ## filter: The :zeek:type:`Log::Filter` instance that controls
    ##         the fate of the given log record.
    type PolicyHook: hook(rec: any, id: ID, filter: Filter);

    ...
}
```

Just as we have patterns for establishing log stream IDs and log-write events, we establish one for creating hooks on log streams. This looks as follows:

```
module Foo;

export {
    # The usual logging ID setup:
    redef enum Log::ID += { LOG };

    ## A default logging policy hook for the stream.
    global log_policy: Log::PolicyHook;
}

...

event zeek_init()
    {
    Log::create_stream(Foo::LOG, [$columns=Info, ..., $policy=log_policy]);
    }
```

These default policy hooks apply at log stream granularity. They are inherited down to log filters. This allows us to express policies across log filters easily. Log filters are free to override those hooks when required. Log filters can also establish policy hooks if none are defined at the stream level. Since these hooks run per log record, they need to be fast -- we don't want to have to string-compare, say, the `filter$name` for every log write.

To implement policy, you just add a hook handler, as follows.

```
hook Foo::log_policy(rec: Foo::Info, id: Log::ID, filter: Log::Filter)
    {
    if ( some_condition_that_vetoes() )
        break;
    }
```

If you want a different (or empty) policy on a given log filter, you can adapt it as usual:

```
hook my_policy(rec: Foo::Info, id: Log::ID, filter: Log::Filter)
    {
    # Can implement arbitrarily here, and also call the default Foo::log_policy hook if needed.
    }

event zeek_init()
    {
    local filter = Log::get_filter(Foo::LOG, "default");
    filter$policy = my_policy;
    Log::add_filter(Test::LOG_OTHER, filter);
    }
```

When a policy hook has no handlers, the logging manager makes sure not to invoke it at all, so there's no overhead -- just as there was no overhead when `$pred` wasn't defined. This is a bit unusual since `Func->Invoke()` already does little in case of unhandled hooks, so let me know if you have concerns here.

One area this PR does not yet address is how to treat alterations to log records. I feel these should be supported. At the moment hooks are free to modify the log record, but this modification "sticks": it remains in effect also in other log filters that run later. That's because the `RecordVal` containing the values currently is a single instance sticking around while all of the filters get processed.

It would be more intuitive for log filters to "own" the record, so when one modifies it, only the write resulting for this filter has the modification. I haven't currently implemented this, but I think one way to do this without doing lots of copying could be this:

- We already make one copy of the record in src/logging/Manager.cc's `Manager::write()`, here:

```
auto columns = columns_arg->CoerceTo({NewRef{}, stream->columns});
```

- Since `RecordVal` supports the notification framework, we could have the log Manager be notified when a hook actually modifies the record, and in that case make sure that the next filter obtains a new copy of the original `columns_arg` passed into `Manager::write()`

This would basically be "copy on write" since we'd make extra copies only when actually required.

Let me know your thoughts...

### CODE REVIEW

File: testing/btest/scripts/base/frameworks/logging/hooks.zeek [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+event zeek_init()
+	{
+	Log::write(Test::LOG, [$t=network_time(), $status="foo"]);
+	}
+
+@TEST-START-NEXT
+
+# Verify that the stream-level policy gets inherited into additional
+# filters.  The single hook handler should get invoked for both of the
+# log filters., and alters them depending on the filter.
```
Comment (rsmmr) [inline testing/btest/scripts/base/frameworks/logging/hooks.zeek:110]:
```suggestion
# log filters, and alters them depending on the filter.
```

**Review by rsmmr — APPROVED** [review]
Looks great.

One the question of modifying records, a couple things:

1. Conceptually, this makes me cringe a little bit because I would  prefer to move away from relying on modifications to event arguments--and this moves the opposite direction.  :-) However,, the logging framework is relying heavily on this already, and seems there's no real harm in adding more oft the same within that context. If we ever wanted to ban mutable event parameters, we'd need to redesign logging anyways. 

2. It may not be worth the trouble of fix the fuzzy semantics of making changes. We are having the same issue generally with events/hooks; in some sense it would be less consistent to fix it here, but not elsewhere. Also not sure that in practice people would even run into this much at all: it would require having multiple filters that want to log different content for the same stream.

So I'd be fine just leaving this as is.

**Comment by JustinAzoff** [comment]
> At the moment hooks are free to modify the log record, but this modification "sticks": it remains in effect also in other log filters that run later. 

This seems like a feature to me.. I'd expect that if you have 2 hooks that modify a log record, that the modifications would stack on top of each other.

```
#redact-field.zeek
hook Foo::log_policy(rec: Foo::Info, id: Log::ID, filter: Log::Filter)
    {
    if (/sensitive/ in rec$field)
        rec$field = "redacted";
    }

#redact-other-field.zeek
hook Foo::log_policy(rec: Foo::Info, id: Log::ID, filter: Log::Filter)
    {
    if (Site::is_local_addr(rec$id$orig_h) && /sensitive/ in rec$other_field)
        rec$other_field = "redacted for local hosts";
    }

#results in potentially both field and other_field redacted

```

**Comment by ckreibich** [comment]
@JustinAzoff, right, within hooks this seems fine and you can control the ordering via priorities if you need that level of control. But the modification also sticks across filters, whose order you can't really control, and those filters might use different hooks. So at that point the notion becomes a bit fuzzy ... but I agree with Robin's point that the whole scenario (conflicting modifications to a field) isn't very likely, and similar conflicts exist elsehwere. So I think we're okay as-is, but should explain it in the upcoming docs for this.

@rsmmr I had completely forgotten the whole line of thought about not modifying the input arguments! What happened to Zeeklets? :-)

**Comment by rsmmr** [comment]
> @rsmmr I had completely forgotten the whole line of thought about not modifying the input arguments! What happened to Zeeklets? :-)

Still on the horizon, but rather far out becauese several pieces need to fall in place first.

File: testing/btest/Baseline/scripts.base.frameworks.intel.remove-non-existing/output [hunk 2]
```diff
@@ -3,9 +3,9 @@
...
 #unset_field	-
 #path	reporter
-#open	2019-03-24-20-20-10
+#open	2020-09-24-00-19-13
 #fields	ts	level	message	location
 #types	time	enum	string	string
-0.000000	Reporter::INFO	Tried to remove non-existing item '192.168.1.1' (Intel::ADDR).	/home/jgras/devel/zeek/scripts/base/frameworks/intel/./main.zeek, lines 563-564
-0.000000	Reporter::INFO	received termination signal	(empty)
-#close	2019-03-24-20-20-10
+1600906753.185591	Reporter::INFO	Tried to remove non-existing item '192.168.1.1' (Intel::ADDR).	/home/christian/devel/zeek/zeek/scripts/base/frameworks/intel/./main.zeek, lines 565-566
```
[the remark this answers is not in the document]
Comment (ckreibich) [reply testing/btest/Baseline/scripts.base.frameworks.intel.remove-non-existing/output:9]: I've not tracked down the exact change, but it did happen a long time ago. I can reproduce the zero-stamps with 2.6.4, and get actual timestamps with any version from 3.0.0 onward.

File: src/logging/Manager.cc [hunk 3]
```diff
@@ -721,17 +730,27 @@ bool Manager::Write(EnumVal* id, RecordVal* columns_arg)
 		Filter* filter = *i;
 		string path = filter->path;
 
+		// Policy hooks may alter the log record or veto the
+		// logging. Optimization: when no hook handlers are
+		// defined, don't even invoke.
+		if ( filter->policy && filter->policy->HasBodies() )
```
[the remark this answers is not in the document]
Comment (rsmmr) [reply src/logging/Manager.cc:736]: So I'll go ahead and suggest we remove the `if`, and can measure after the merge to make sure. I wouldn't expect a noticeable difference.

[the remark this answers is not in the document]
Comment (rsmmr) [reply src/logging/Manager.cc:736]: @christian I can do that change during merge if that sounds good to you.

[the remark this answers is not in the document]
Comment (ckreibich) [reply src/logging/Manager.cc]: Robin I took a few measurements. I wrote log entries in a tight loop, in three different scenarios: no hooks in the picture at all (via streams that don't assign to `$policy`), using the default hook design with no hook handlers, and with a single no-op hook handler. The overhead of invoking the hook when it has no handlers is measurable but at most a few percentage points. So yeah, I agree — let's not optimize now, and see first whether it matters enough. I've pushed an update that takes out the optimization, but gives some context in a comment.

**Comment by ckreibich** [comment]
I'm putting together a docs update for this, PR to follow.

**Comment by rsmmr** [comment]
Is this ready to go?

**Pull request merged**